src/validation/rules/SingleFieldSubscriptionsRule.js

```diff
@@ -1,28 +1,119 @@
 import { GraphQLError } from '../../error/GraphQLError';
 
 import type { ASTVisitor } from '../../language/visitor';
-import type { OperationDefinitionNode } from '../../language/ast';
+import type {
+  OperationDefinitionNode,
+  SelectionSetNode,
+} from '../../language/ast';
+import { Kind } from '../../language/kinds';
 
 import type { ASTValidationContext } from '../ValidationContext';
 
 /**
- * Subscriptions must only include one field.
+ * Walks the selection set and returns a list of selections where extra fields
+ * were selected, and selections where introspection fields were selected.
+ */
+function walkSubscriptionSelectionSet(
+  context: ASTValidationContext,
+  selectionSet: SelectionSetNode,
+  responseKeys,
+  fieldNames = new Set(),
+  visitedFragmentNames = new Set(),
+  extraFieldSelections = [],
+  introspectionFieldSelections = [],
+) {
+  for (const selection of selectionSet.selections) {
+    switch (selection.kind) {
+      case Kind.FIELD: {
+        const fieldName = selection.name.value;
+        fieldNames.add(fieldName);
+        const responseName = selection.alias
+          ? selection.alias.value
+          : fieldName;
+        responseKeys.add(responseName);
+        if (fieldName[0] === '_' && fieldName[1] === '_') {
+          // fieldName represents an introspection field if it starts with `__`
+          introspectionFieldSelections.push(selection);
+        } else if (fieldNames.size > 1 || responseKeys.size > 1) {
+          extraFieldSelections.push(selection);
+        }
```

Comment:
I think this is the only bit that's unique from `CollectFields()` - I'd suggest moving this bit of logic down to the Operation visit function after the grouped fields are collected.

Reply:
Have instead delegated to `collectFields`